File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        left, right = 0, len(height) - 1
        left_max, right_max = 0, 0
        result = 0

        while left < right:
            left_max = max(left_max, height[left])
            right_max = max(right_max, height[right])

            if height[left] < height[right]:
                result += left_max - height[left]
                left += 1
            else:
                result += right_max - height[right]
                right -= 1

        return result

File: 42-trapping-rain-water/trapping-rain-water.c
int trap(int* height, int heightSize) {
    int left = 0, right = heightSize - 1;
    int left_max = 0, right_max = 0;
    int result = 0;

    while (left < right) {
        left_max = height[left] > left_max ? height[left] : left_max;
        right_max = height[right] > right_max ? height[right] : right_max;

        if (height[left] < height[right]) {
            result += left_max - height[left];
            left++;
        } else {
            result += right_max - height[right];
            right--;
        }
    }

    return result;
}